File: 438-find-all-anagrams-in-a-string/find-all-anagrams-in-a-string.py
class Solution(object):
    def findAnagrams(self, s, p):
        """
        :type s: str
        :type p: str
        :rtype: List[int]
        """
        need={}
        window={}
        result=[]

        #update need map
        for ch in p:
            need[ch]=need.get(ch,0)+1

        #update window map
        left=0
        for right in range(len(s)):
            ch=s[right]
            window[ch]=window.get(ch,0)+1
            
            #check window length exceeds condition then move right
            if right-left+1 > len(p):
                leftchar=s[left]
                window[leftchar]-=1
                if window[leftchar]==0:
                    del window[leftchar]
                left+=1

            #if matching then append left to result
            if window == need:
                result.append(left)

        return result

File: 438-find-all-anagrams-in-a-string/find-all-anagrams-in-a-string.java
class Solution {
    public List<Integer> findAnagrams(String s, String p) {
        HashMap<Character,Integer> mapP = new HashMap<>();
        HashMap<Character,Integer> mapW = new HashMap<>();

        //add p into mapP
        for (int i=0;i<p.length();i++){
            mapP.put(p.charAt(i),mapP.getOrDefault(p.charAt(i),0)+1);
        }
        //declare variables
        int left=0;
        int right=0;
        List<Integer> result = new ArrayList<>(); //syntax doubt

        while(right<s.length()){
            //update window
            mapW.put(s.charAt(right),mapW.getOrDefault(s.charAt(right),0)+1);

            //window length check
            if((right-left+1)>p.length()){
                char ch = s.charAt(left);
                mapW.put(ch,mapW.get(ch)-1);
                if(mapW.get(ch)==0){
                    mapW.remove(ch);
                }
                left++;
            }
            //check frequencies match
            if((right-left+1)==p.length()){
                if (mapP.equals(mapW)){
                    result.add(left);
                }
            }
            right++;
        }
        return result;
    }
}